vpr/src/device/rr_spatial_lookup.h

```diff
@@ -0,0 +1,109 @@
+class RRSpatialLookup {
+    /* -- Constructors -- */
+  public:
+    /* Explicitly define the only way to create an object */
+    explicit RRSpatialLookup(t_rr_node_indices& rr_node_indices);
+
+    /* Disable copy constructors and copy assignment operator
+     * This is to avoid accidental copy because it could be an expensive operation considering that the 
+     * memory footprint of the data structure could ~ Gb
+     * Using the following syntax, we prohibit accidental 'pass-by-value' which can be immediately caught 
+     * by compiler
+     */
+    RRSpatialLookup(const RRSpatialLookup&) = delete;
+    void operator=(const RRSpatialLookup&) = delete;
+
+    /* -- Accessors -- */
+  public:
+    /* Returns the index of the specified routing resource node.  
+     * - (x, y) are the grid location within the FPGA
+     * - rr_type specifies the type of resource,
+     * - ptc gives a unique number of resources of that type (e.g. CHANX) at that (x,y).
+     *   All ptcs start at 0 and are positive.
+     *   Depending on what type of resource this is, ptc can be 
+     *     - the class number of a common SINK/SOURCE node of grid, 
+     *       starting at 0 and go up to class_inf size - 1 of SOURCEs + SINKs in a grid
+     *     - pin number of an input/output pin of a grid. They would normally start at 0
+     *       and go to the number of pins on a block at that (x, y) location
+     *     - track number of a routing wire in a channel. They would normally go from 0
+     *       to channel_width - 1 at that (x,y)
+     *
+     * Note that for segments (CHANX and CHANY) of length > 1, the segment is
+     * given an rr_index based on the (x,y) location at which it starts (i.e.
+     * lowest (x,y) location at which this segment exists).
+     *
+     * The 'side' argument only applies to IPIN/OPIN types, and specifies which
+     * side of the grid tile the node should be located on. The value is ignored
+     * for non-IPIN/OPIN types
+     *
+     * This routine also performs error checking to make sure the node in
+     * question exists.
+     */
+    RRNodeId find_node(int x,
+                       int y,
+                       t_rr_type type,
+                       int ptc,
+                       e_side side = NUM_SIDES) const;
+
+    /* -- Mutators -- */
+  public:
+    /* Register a node in the fast look-up 
+     * - You must have a valid node id to register the node in the lookup
+     * - (x, y) are the coordinate of the node to be indexable in the fast look-up
+     * - type is the type of a node
+     * - ptc is a feature number of a node, which can be
+     *   - the class number of a common SINK/SOURCE node of grid, 
+     *   - pin index in a tile when type is OPIN/IPIN
+     *   - track index in a routing channel when type is CHANX/CHANY
+     * - side is the side of node on the tile, applicable to OPIN/IPIN 
+     *
+     * An invalid id will be returned if the node does not exist
```

Comment:
I think this comment (an invalid id will be returned if the node does not exist) is in the wrong spot.  It should be with find_node, not add_node.

Reply:
You are right. I moved to the find_node() API.